src/solvers/quasi_Newton.jl

```diff
@@ -345,6 +345,11 @@ function step_solver!(mp::AbstractManoptProblem, qns::QuasiNewtonState, iter)
     M = get_manifold(mp)
     get_gradient!(mp, qns.X, qns.p)
     qns.direction_update(qns.η, mp, qns)
+    if real(inner(M, qns.p, qns.η, qns.X)) > 0
```

Comment:
Much more than a rounding error. You can try running the `test_case_manopt()` function from the PR to observe the issue. The direction gets wrong by the third iteration. I'd prefer a better fix too but it's fairly difficult to figure it out.

It never appears in my tests on flat Euclidean where Manopt's iterations are very close to Optim.jl's iterations. So, I don't know, but maybe it has something to do with vector transports. Both projection transport and parallel transport give this error though.

Reply:
With this fix, both Manopt and Optim converge well on Rosenbrock with spherical constraints. Sometimes Manopt is faster, and sometimes Optim. For Euclidean Rosenbrock, Manopt is consistently a bit faster.

Reply:
Well, this “fix” is till of the kind “if it is wrong reset, restart” instead of finding out what is wrong. For projection transport I think this _can_ happen, since that only approximates, it should not happen for parallel transport. But with now 3 master students, I will not have much time beyond my own small PRs I have planned (or want to finish). 


But I would still prefer to not have such duck-typing fixes in the code. If it is wrong, let's better find the error than just saying “Error appeared? Restart!”

Reply:
Or more precisely, you are fixing a symptom, but not its origin. Sure, fixing the origin takes more time, but I see no reason to rush here?
(If your car breaks a tire, an actual working new tire would also be better than a large spool of duct tape wrapped around the old tire and saying: Look! Nearly a tire again – and on even grounds, still reasonably fast!)

Reply:
Is the search direction in Riemannian L-BFGS actually guaranteed to be a descent direction from theory?

> Sure, fixing the origin takes more time, but I see no reason to rush here?

People are actually driving a car with a flat tire right now by using Manopt, so duct-taping it seems better than not doing anything?

Reply:
For the first, I will check but I am quite sure (among others because the Matlab version has, to the best of my knowledge not such a duct tape fix).

You are right duct taping is maybe better than not fixing it, but we should also be careful not to forget about it then, which I see as a danger here if the leave it as you propose here. I would then _at least_ add a warning.

But I can try to find time to take a look at this as well, just not this week, since it is already now super-packed.

Reply:
Sure, a warning and for example an open issue to remind us to look into it later would be fine. Anyway, I'm not sure we can 100% guarantee that there won't be an incorrect direction there, and having that check even after fixing the underlying issue might still be a good precaution. Whether it should restart or throw an error, I'm not sure. Restarting means getting an answer but more slowly, throwing means no answer at all.

Reply:
I do not want to throw an error, that is for sure ;)

If you want to keep the check, then please with a warning that states what occurred and what is done to fix it.

And we should also open an issue with the smallest example you can generate :)

Reply:
Hm, it seems to only happen with my modification of HZ. I will keep an eye for a more stable reproducer.

Reply:
Oh, that sounds super interesting, but would maybe mean, you have a bug then? If it is not for the reason of qN itself, I would again prefer not to have this warning here;
but one could indeed come up with a “VerificationState” (similar to the same manifold) where in each iteration one allows for a certain test(s) to be performed.

Reply:
I'm trying to figure it out but a bug in linesearch doesn't sound like a plausible explanation to me.

`VerificationState` sounds like something that would be nightmarishly difficult to implement. I believe such optional sanity checks could be useful in the future, even if they are off by default.

Reply:
Ha, I've found an independent reproducer: #346 .

Reply:
Oh I did not want to imply _all_ linesearch is wrong but maybe your improvements had a slight flaw in one specific case. But great that you found an example and good that it even works with PT.
Then we could indeed keep the warning here and keep that (and the duct tape fix) until we fix the bug.

Reply:
I think we discussed this enough in Zulip, and dressed this in tutorial mode. I hope I even added it here? We could add a tutorial debug for this if you feel that helps.

Reply:
Hm, good that you remind me. I thought we already had this check merged in Manopt.jl? This is 100% not a tutorial-only issue because some combinations of line search/manifold/vector transport will lead to non-descent directions here which we need to catch (or the user gets a very hard to debug problem). Those combinations are still useful because they sometimes converge faster than "always descent direction" alternatives.